### PR TITLE
[SE-3950] Add a waffle flag to ignore the accept-language header

### PR DESCRIPTION
**Pull request by Kelketek**
This pull request adds a piece of middleware, controlled via waffle flag, to remove the Accept-Language header from a request so that LANGUAGE_CODE is set as the default language in the event no cookie is set.

**JIRA tickets**: https://tasks.opencraft.com/browse/SE-3950

**Testing instructions**:

1. Clone the repository if you have not done so already.
2. Create a virtualenv with Python 2.7 and activate it.
3. Follow the installation instructions in the README.
4. Set ALLOWED_HOSTS to ['*'] to make things easier on yourself while testing.
4. Change the LANGUAGE_CODE setting to a language you'd prefer.
5. Use runserver. Observe that the language header is still respected.
6. Set the waffle flag with `./manage.py waffle_flag ignore_accept_language --everyone --create`
7. Reload the page and verify the language has changed.

**Author notes and concerns**:

Up until this point, Campus has not customized insights. They were running off of edx's `open-release/ginkgo-master` branch. This changes that to run off of this repository here, and so that will need to be reflected in configuration changes.

We may be able to upstream a change like this, but the code will almost certainly be different-- the way Middleware is handled by Django upstream changed significantly several versions since.

**Reviewers**
- [ ] @nizarmah

### CODE REVIEW

File: analytics_dashboard/core/middleware.py
Comment (nizarmah) [inline analytics_dashboard/core/middleware.py:16]: @Kelketek even though the solution looks good, I can't accept it because it would be introducing additional code drift to _edx-olive_. 

We'll need an upstream pull request to be created in order to upstream the changes to `master`. And the pull request to _edx-olive_ should be a backport for that change.

File: analytics_dashboard/core/middleware.py
Comment (nizarmah) [inline analytics_dashboard/core/middleware.py]: I would prefer if you use a better name for the waffle switch/flag, especially since we're going to be upstreaming this middleware. Something of the sorts of `analytics_dashboard.use_default_language` or `analytics_dashboard.ignore_browser_language`.

Also, when defining the Waffle Switch/Flag in the `config.py` or `config/waffle.py`, please use [this template](https://edx.readthedocs.io/projects/edx-toggles/en/latest/how_to/documenting_new_feature_toggles.html#boilerplate-template) to document it.

File: analytics_dashboard/core/middleware.py
Comment (nizarmah) [inline analytics_dashboard/core/middleware.py:11]: Since we do want the solution to be upstreamed, I'm going to have to be a bit picky about this change, sorry in advance.

I would prefer if we separate any waffle functionality from here, and move the Waffle Switch/Flag creation to `analytics_dashboard/core/config`. The config can be a directory which contains `waffle.py` or it can be a python file which contains the defined waffle switch/flag. 

And instead of checking `flag_is_active` here, we'd import the Waffle Switch/Flag, and use `SWITCH_VARIABLE.is_enabled()` to check if it is enabled or not.

**Review by nizarmah — CHANGES_REQUESTED** [review]
@Kelketek I left some comments on this pull request. Once those comments are addressed, please create a pull request to [edx/edx-analytics-dashboard](https://github.com/edx/edx-analytics-dashboard).

File: analytics_dashboard/core/middleware.py
Comment (nizarmah) [inline analytics_dashboard/core/middleware.py:25]: Small nit pick: new line before this line please

```suggestion

        if request.META.has_key('HTTP_ACCEPT_LANGUAGE'):
```

File: analytics_dashboard/core/middleware.py
Comment (Kelketek) [reply analytics_dashboard/core/middleware.py:16]: We cannot upstream this code in this fashion because edx-insights upstream uses a more recent version of Django, which changes how middleware works. Also upstream uses the Django plugin utils. The purpose of this PR is to get the changes available in a way that Campus can use today. We will then make a follow-up PR to add the changes to upstream in a way that's compatible with the newer versions of Django and which passes edX's requirements.

We do not, at least last I checked, wait for the upstream version to be merged before getting the client's instance patched. Instead, we fix the client version, then upstream the changes, which can take several weeks. Once the changes are upstreamed, we either make a commit to remove the changes we made and update to the latest upstream version, or we wait until the next rebase and remove the commit as part of the process.

The comment above makes it easy for a person doing such a rebase to verify if we got upstream to merge the change in time.

File: analytics_dashboard/core/middleware.py
Comment (Kelketek) [reply analytics_dashboard/core/middleware.py]: Thanks for showing me that boilerplate. I was searching around looking for something like that and couldn't find anything.

File: analytics_dashboard/core/middleware.py
Comment (nizarmah) [reply analytics_dashboard/core/middleware.py:16]: > The purpose of this PR is to get the changes available in a way that Campus can use today. We will then make a follow-up PR to add the changes to upstream in a way that's compatible with the newer versions of Django and which passes edX's requirements.

That sounds reasonable, as long as it is done in the scope of the same ticket.
As long as there are changes which can be backported to a different release, we can create such a Pull Request for `edx-olive` for now.

But it should be indicated in the pull requested that this change should be dropped once we upgrade them to a new release. And we'll need to indicate which pull request we'll need to backport when upgrading them to a new release. 

> We do not, at least last I checked, wait for the upstream version to be merged before getting the client's instance patched. Instead, we fix the client version, then upstream the changes, which can take several weeks. Once the changes are upstreamed, we either make a commit to remove the changes we made and update to the latest upstream version, or we wait until the next rebase and remove the commit as part of the process.

You're absolutely right, I'm sorry for not explaining myself properly. What I was referring to is preparing the changes for `master` and then backporting them to the specific release, before opening the pull request. 

Then the changes get reviewed internally on the backported pull request, once approved, any changes requested there get applied to the upstream pull request, and then the upstream pull request gets reviewed by an internal member. 

But what I was referring to mainly is backporting the changes that were applied to `master`.

However, if we do as you mentioned, and apply something temporary which we'd throw away later on, then we can consider making two separate changes. However, I'd like the opinions of @itsjeyd and/or @clemente regarding this.

**Comment by Kelketek** [comment]
@nizarmah I've modified this pull request to use the same waffle flag name as convention requires, and which is now in the upstream PR.

I didn't end up adding edx-toggles to this version or adding the relevent boilerplate because edx-toggles hasn't supported 2.7 for a long time. As mentioned earlier, it should be easy to remove this commit when rebasing later, especially since the code is so clearly marked.

File: analytics_dashboard/core/middleware.py
Comment (clemente) [reply analytics_dashboard/core/middleware.py:16]: @Kelketek @nizarmah 

> We do not, at least last I checked, wait for the upstream version to be merged before getting the client's instance patched. 

This depends on each client. With some we do {upstream fix + client backport} at the same time. With Campus, from what I know from @itsjeyd, we *do* wait for the fix to be merged upstream. Some time ago I added this particular information to the epic description (SE-311), please check the `Upstreaming strategy` section there.
However, for urgent changes we may do the extra work of backporting. I don't know who decides when to do it (CC @itsjeyd).
I like always backporting; otherwise the client must wait for months. But we may need to follow the standard Campus approach. We can keep the general upstreaming discussion at SE-311.

Comment (Kelketek) [reply analytics_dashboard/core/middleware.py:16]: Thanks for the clarity here, @clemente . I actually need to close this PR, though, since as discussed in the ticket we didn't end up needing it anyway. :/ We can continue discussion about the upstream PR there, though, since it still has some useful attributes.

**Pull request closed without merging**
